File: C++/CocktailSort.cpp
// C++ implementation of Cocktail Sort
#include <iostream>
#include <time.h>
#include <stdlib.h>
#include <chrono>
using namespace std;

// Sorts array a[0..n-1] using Cocktail sort
void CocktailSort(int a[], int n)
{
	bool swapped = true;
	int start = 0;
	int end = n - 1;

	while (swapped)
	{
		// reset the swapped flag on entering
		// the loop, because it might be true from
		// a previous iteration.
		swapped = false;

		// loop from left to right same as
		// the bubble sort
		for (int i = start; i < end; ++i)
		{
			if (a[i] > a[i + 1]) {
				swap(a[i], a[i + 1]);
				swapped = true;
			}
		}

		// if nothing moved, then array is sorted.
		if (!swapped)
			break;

		// otherwise, reset the swapped flag so that it
		// can be used in the next stage
		swapped = false;

		// move the end point back by one, because
		// item at the end is in its rightful spot
		--end;

		// from right to left, doing the
		// same comparison as in the previous stage
		for (int i = end - 1; i >= start; --i)
		{
			if (a[i] > a[i + 1]) {
				swap(a[i], a[i + 1]);
				swapped = true;
			}
		}

		// increase the starting point, because
		// the last stage would have moved the next
		// smallest number to its rightful spot.
		++start;
	}
}

/* Prints the array */
void printArray(int a[], int n)
{
	for (int i = 0; i < n; i++)
		printf("%d ", a[i]);
	printf("\n");
}


// Driver code
int main()
{	
	// array 1
	int arr[100];
 
	srand(time(NULL));
 
	for(int i=0;i<100;i++)
	{
		int num = rand() %1000 ;
		arr[i]=num;
	}

	int n = sizeof(arr) / sizeof(arr[0]);

	chrono::steady_clock::time_point begin = chrono::steady_clock::now();
	CocktailSort(arr, n);
	chrono::steady_clock::time_point end = chrono::steady_clock::now();
	printf("Sorted array :\n");
	printArray(arr, n);
	cout << "Time difference = " << chrono::duration_cast<chrono::microseconds>(end - begin).count() << "[µs]" << std::endl;


	// array 2

	int arr2[1000];

	for(int i=0;i<1000;i++)
	{
		int num = rand() %1000 ;
		arr2[i]=num;
	}

	int n2 = sizeof(arr2) / sizeof(arr2[0]);

	chrono::steady_clock::time_point begin2 = chrono::steady_clock::now();
	CocktailSort(arr2, n2);
	chrono::steady_clock::time_point end2 = chrono::steady_clock::now();
	printf("Sorted array :\n");
	printArray(arr2, n2);
	cout << "Time difference = " << chrono::duration_cast<chrono::microseconds>(end2 - begin2).count() << "[µs]" << std::endl;

	// array 3

	int arr3[10000];

	for(int i=0;i<10000;i++)
	{
		int num = rand() %1000 ;
		arr3[i]=num;
	}

	int n3 = sizeof(arr3) / sizeof(arr3[0]);

	chrono::steady_clock::time_point begin3 = chrono::steady_clock::now();
	CocktailSort(arr3, n3);
	chrono::steady_clock::time_point end3 = chrono::steady_clock::now();
	printf("Sorted array :\n");
	printArray(arr3, n3);
	cout << "Time difference = " << chrono::duration_cast<chrono::microseconds>(end3 - begin3).count() << "[µs]" << std::endl;

	return 0;
}

File: Python/CocktailSort.py
# Python program for implementation of Cocktail Sort
from datetime import datetime
import numpy as np


def cocktailSort(a):
	n = len(a)
	swapped = True
	start = 0
	end = n-1
	while (swapped == True):

		# reset the swapped flag on entering the loop,
		# because it might be true from a previous
		# iteration.
		swapped = False

		# loop from left to right same as the bubble
		# sort
		for i in range(start, end):
			if (a[i] > a[i + 1]):
				a[i], a[i + 1] = a[i + 1], a[i]
				swapped = True

		# if nothing moved, then array is sorted.
		if (swapped == False):
			break

		# otherwise, reset the swapped flag so that it
		# can be used in the next stage
		swapped = False

		# move the end point back by one, because
		# item at the end is in its rightful spot
		end = end-1

		# from right to left, doing the same
		# comparison as in the previous stage
		for i in range(end-1, start-1, -1):
			if (a[i] > a[i + 1]):
				a[i], a[i + 1] = a[i + 1], a[i]
				swapped = True

		# increase the starting point, because
		# the last stage would have moved the next
		# smallest number to its rightful spot.
		start = start + 1


# Random integer numpy array
a = np.random.randint(1000, size = 100)
b = np.random.randint(1000, size = 1000)
c = np.random.randint(1000, size = 10000)

# Sort array a
start_time = datetime.now()

cocktailSort(a)

end_time = datetime.now()

print("Sorted array is:")
for i in range(len(a)):
	print("% d" % a[i])

a1 = (end_time - start_time)

# Sort array b
start_time = datetime.now()

cocktailSort(b)

end_time = datetime.now()

print("Sorted array is:")
for i in range(len(b)):
	print("% d" % b[i])

b1 = (end_time - start_time)

#Sort array c
start_time = datetime.now()

cocktailSort(c)

end_time = datetime.now()

print("Sorted array is:")
for i in range(len(c)):
	print("% d" % c[i])

c1 = (end_time - start_time)

print("Run time  : a = ", a1, " b = ", b1, " c = ", c1)

